// react-chunk/lib/Recorder.js
'use strict';

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ChunkRecorder = function (_React$Component) {
  _inherits(ChunkRecorder, _React$Component);

  function ChunkRecorder() {
    _classCallCheck(this, ChunkRecorder);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  ChunkRecorder.prototype.getChildContext = function getChildContext() {
    return {
      chunks: {
        addChunk: this.props.addChunk
      }
    };
  };

  ChunkRecorder.prototype.render = function render() {
    return _react2.default.Children.only(this.props.children);
  };

  return ChunkRecorder;
}(_react2.default.Component);

ChunkRecorder.propTypes = {
  addChunk: _propTypes2.default.func.isRequired,
  children: _propTypes2.default.oneOfType([_propTypes2.default.node, _propTypes2.default.func]).isRequired
};
ChunkRecorder.childContextTypes = {
  chunks: _propTypes2.default.shape({
    addChunk: _propTypes2.default.func
  })
};


module.exports = ChunkRecorder;
